libraries/networking/src/AssetClient.cpp

```diff
@@ -0,0 +1,208 @@
+        packet->writePrimitive(messageID);
+        packet->write(hash.toLatin1().constData(), 32);
+        packet->writePrimitive(start);
+        packet->writePrimitive(end);
+
+        nodeList->sendPacket(std::move(packet), *assetServer);
+
+        _pendingRequests[messageID] = callback;
+
+        return true;
+    }
+
+    return false;
+}
+
+bool AssetClient::getAssetInfo(QString hash, GetInfoCallback callback) {
+    auto nodeList = DependencyManager::get<NodeList>();
+    SharedNodePointer assetServer = nodeList->soloNodeOfType(NodeType::AssetServer);
+
+    if (assetServer) {
+        auto packet = NLPacket::create(PacketType::AssetGetInfo);
+
+        auto messageID = ++_currentID;
+        packet->writePrimitive(messageID);
+        packet->write(hash.toLatin1().constData(), 32);
+
+        nodeList->sendPacket(std::move(packet), *assetServer);
+
+        _pendingInfoRequests[messageID] = callback;
+
+        return true;
+    }
+
+    return false;
+}
+
+void AssetClient::handleAssetGetInfoReply(QSharedPointer<NLPacket> packet, SharedNodePointer senderNode) {
+    MessageID messageID;
+    packet->readPrimitive(&messageID);
+    auto assetHash = QString(packet->read(HASH_HEX_LENGTH));
+    
+    AssetServerError error;
+    packet->readPrimitive(&error);
+
+    AssetInfo info;
+
+    if (!error) {
+        packet->readPrimitive(&info.size);
+    }
+
+    if (_pendingInfoRequests.contains(messageID)) {
+        auto callback = _pendingInfoRequests.take(messageID);
+        callback(error != NO_ERROR, info);
+    }
+}
+
+void AssetClient::handleAssetGetReply(QSharedPointer<NLPacketList> packetList, SharedNodePointer senderNode) {
+    QByteArray data = packetList->getMessage();
+    QBuffer packet { &data };
+    packet.open(QIODevice::ReadOnly);
+
+    auto assetHash = packet.read(HASH_HEX_LENGTH);
+    qDebug() << "Got reply for asset: " << assetHash;
+
+    MessageID messageID;
+    packet.read(reinterpret_cast<char*>(&messageID), sizeof(messageID));
+
+    AssetServerError error;
+    packet.read(reinterpret_cast<char*>(&error), sizeof(AssetServerError));
+    QByteArray assetData;
+
+    if (!error) {
+        DataOffset length;
+        packet.read(reinterpret_cast<char*>(&length), sizeof(DataOffset));
+        data = packet.read(length);
+    } else {
+        qDebug() << "Failure getting asset: " << error;
+    }
+
+    if (_pendingRequests.contains(messageID)) {
+        auto callback = _pendingRequests.take(messageID);
+        callback(!error, data);
+    }
+}
+
+bool AssetClient::uploadAsset(QByteArray data, QString extension, UploadResultCallback callback) {
+    auto nodeList = DependencyManager::get<NodeList>();
+    SharedNodePointer assetServer = nodeList->soloNodeOfType(NodeType::AssetServer);
+    if (assetServer) {
+        auto packetList = std::unique_ptr<NLPacketList>(new NLPacketList(PacketType::AssetUpload, QByteArray(), true, true));
+
+        auto messageID = _currentID++;
```

Comment:
You use `++_currentID` for all other increments, so this might pose an issue where 2 messageID would be the same.